test/framework/prometheusagent.go

```diff
@@ -0,0 +1,129 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package framework
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	monitoringv1alpha1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1alpha1"
+	"github.com/prometheus-operator/prometheus-operator/pkg/operator"
+	"github.com/prometheus-operator/prometheus-operator/pkg/prometheus/agent"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (f *Framework) MakeBasicPrometheusAgent(ns, name, group string, replicas int32) *monitoringv1alpha1.PrometheusAgent {
+	return &monitoringv1alpha1.PrometheusAgent{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   ns,
+			Annotations: map[string]string{},
+		},
+		Spec: monitoringv1alpha1.PrometheusAgentSpec{
+			CommonPrometheusFields: monitoringv1.CommonPrometheusFields{
+				Replicas: &replicas,
+				Version:  operator.DefaultPrometheusVersion,
+				ServiceMonitorSelector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"group": group,
+					},
+				},
+				PodMonitorSelector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"group": group,
+					},
+				},
+				ServiceAccountName: "prometheus",
+				Resources: v1.ResourceRequirements{
+					Requests: v1.ResourceList{
+						v1.ResourceMemory: resource.MustParse("400Mi"),
+					},
+				},
+			},
+		},
+	}
+}
+
+func (f *Framework) CreatePrometheusAgentAndWaitUntilReady(ctx context.Context, ns string, p *monitoringv1alpha1.PrometheusAgent) (*monitoringv1alpha1.PrometheusAgent, error) {
+	result, err := f.MonClientV1alpha1.PrometheusAgents(ns).Create(ctx, p, metav1.CreateOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("creating %v prometheus-agent instances failed (%v): %v", p.Spec.Replicas, p.Name, err)
+	}
+
+	if err := f.WaitForPrometheusAgentReady(ctx, result, 5*time.Minute); err != nil {
+		return nil, fmt.Errorf("waiting for %v prometheus-agent instances timed out (%v): %v", p.Spec.Replicas, p.Name, err)
+	}
+
+	return result, nil
+}
+
+func (f *Framework) WaitForPrometheusAgentReady(ctx context.Context, p *monitoringv1alpha1.PrometheusAgent, timeout time.Duration) error {
+	expected := *p.Spec.Replicas
+	if p.Spec.Shards != nil && *p.Spec.Shards > 0 {
+		expected = expected * *p.Spec.Shards
+	}
+
+	if err := f.WaitForResourceAvailable(
+		ctx,
+		func(context.Context) (resourceStatus, error) {
+			current, err := f.MonClientV1alpha1.PrometheusAgents(p.Namespace).Get(ctx, p.Name, metav1.GetOptions{})
+			if err != nil {
+				return resourceStatus{}, err
+			}
+			return resourceStatus{
+				expectedReplicas: expected,
+				generation:       current.Generation,
+				replicas:         current.Status.AvailableReplicas,
```

Comment:
the `WaitForPrometheusAgentReady` fuction was copy pasted from the Prometheus test framework and this line specifically was setting `replicas: current.StatusUpdatedReplicas` instead of `AvailableReplicas`.

A month later and I finally realized that UpdatedReplicas = 1 doesn't mean that prometheus-agent is ready and we should be looking at AvailableReplicas instead 😅

Reply:
After a few more hours of debugging, I've found out that https://github.com/prometheus-operator/prometheus-operator/blob/906669420e97573a4ea7534fb3fdf0ce8192ae03/test/framework/status.go#L50-L64 is where the bug lives

![image](https://github.com/prometheus-operator/prometheus-operator/assets/24193764/e2002845-445b-4a2b-939a-0c3a0579260f)

Notice in the image above how the "available" variable is of type "reconciled" instead of available. It is set to `true`, even though kubectl is showing false for the `AVAILABLE` field.

I'm still trying to understand why this is happening, but I suspect it has something to do with the pointer assignments

Reply:
Found it 🥳 That was a really hard one, what the hell

The problem was that we assign the address of `cond` to the variable `available`. On the second loop, the address of `cond` changes, and so does the value of `available`.

I've pushed a fix in 53c708ab8f10ca78b65fe689ac732a38f2f37492, making sure we're not changing values of `available` and `reconciled` on every loop

Reply:
Gosh, that was a month and a half of headaches 😅

Reply:
OK! I think I got it! So in the second loop if the resource is already reconciled this function will return `True` because both `available` and `reconciled` will be pointing to the conditions `reconciled` which is already `true`. Damn, that's indeed a thought one! Great job Arthur!

Reply:
Thanks for that, @ArthurSens! I actually had the same issue trying to create some for ScrapeConfig and didn't understand why it wasn't waiting for the pods to be ready.